### ISSUE TITLE
What is HashSet & Streams

### ISSUE THREAD

**Issue by huiyiiih** [nus-cs2103-AY1718S1/forum#12]
Hello everyone, 

I have read about what Streams means but I still do not really get it. It is stated in tutorialspoint that streams represents a sequence of objects from a source. What does this means? What is pipelining? 
I also do not understand how they are used. Even though there are examples in the website. But I am still unsure how they are used and why.

What does HashSet exactly do? 
I do not understand what is this line doing.

<img width="659" alt="screen shot 2017-08-29 at 5 24 39 pm" src="https://user-images.githubusercontent.com/31040589/29818125-78108fca-8ced-11e7-931c-99afc5d896bd.png">

**Comment by damithc** [comment]
@huiyiiih I'll wait for someone to answer the 2nd question, which is not related to streams at all. 

Streams takes some time to understand, but you don't need to know streams for addressbook-level1. I hope to provide some resources about streams in a later week.

**Comment by damithc** [comment]
> What does HashSet exactly do?

It is a `Set` type data structure. Keep in mind that the main feature of a `Set` is that it does not allow duplicates. See if this helps https://beginnersbook.com/2013/12/hashset-class-in-java-with-example/

**Comment by Xenonym** [comment]
A Set in Java is the same as a mathematical set: it stores unique elements (ie. there should not be any duplicates).  A HashSet is basically a set that is implemented by a hashtable (remember CS1020?).

What that line is doing can be broken down into a few steps. Imagine that we have a `person` named `"John Lim Say Lim"`:
1. `getNameFromPerson(person)` will get the name `"John Lim Say Lim"`
2. `splitByWhitespace` splits the name by spaces and returns a list of words, yielding an ArrayList with the strings `{"John", "Lim", "Say", "Lim"}`.
3. `new HashSet<>` will create a new HashSet of the above ArrayList, *with duplicate elements removed*, yielding `{"John, "Lim", "Say"}`.

TL;DR - the line that you refer to basically gets a list of words in the person's name without duplicate words!

**Comment by huiyiiih** [comment]
@damithc @Xenonym Thank you so much.

I tried doing it for the getEmailFromPerson and getPhoneFromPerson but it doesn't work. Is this because I can't do HashSet with only one string? 

The error was "Unchecked assignment: 'java.util.HashSet' to 'java.util.Set<java.lang.String>'" and "java: cannot infer type arguments for java.util.HashSet<>".

**Comment by Xenonym** [comment]
@huiyiiih sort of. To create a HashSet with existing elements, you have to pass in a `Collection` eg. `ArrayList, LinkedList` of elements, not just a single element by itself. In your case, passing in the email or phone number in an `ArrayList`  (with just the phone number or email inside) should work. This will create a `HashSet` with one element (the phone number or email).

**Comment by bokwoon95** [comment]
The easiest way I picture streams is like math function chaining:

> `(e∘f∘g)(x) = (e(f(g(x))))`
> x gets passed into g(), then f(), then e()

where the output of one function gets passed as input to another function, and so on. That is essentially what 'pipelining' is.

But instead of 1 variable you have a stream of variables, they all pass through different functions that each do different things on it. For example a stream could look like this
>start: [a, b, c, e, f, g] 
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-> (convert to uppercase) 
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;->[A, B, C, D, E, F, G]
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-> (remove vowels)
>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-> [B, C, D, F, G]: end

In Java it could look something like (not accurate, but has the general idea):
> output_array =
>array.stream().map(s -> s.toUpperCase()).filter(s -> !s.isVowel()).list).collect(Collectors.toArray)

or more legibly

> output_array = 
> array
>&nbsp;&nbsp;&nbsp;&nbsp;.stream()&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//converts the array into a stream first
>&nbsp;&nbsp;&nbsp;&nbsp;.map(s -> s.toUpperCase())&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//for each item in the stream, convert it to Uppercase
>&nbsp;&nbsp;&nbsp;&nbsp;.filter(s -> !s.isVowel())&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//for each item that is a consonant, keep it inside the stream
>&nbsp;&nbsp;&nbsp;&nbsp;.collect(Collectors.toArray())&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;//converts the stream back into an array

PS: The `s -> s.toUpperCase()` notation is Java's anonymous function notation, which is basically means given an item `s`, perform the action to the right of the arrow sign `->` on `s`.

**Comment by damithc** [comment]
Closing because this seems to be resolved. Please reopen if still not resolved.
